modules/ephemeris/src/test/scala/gem/horizons/EphemerisParserSpec.scala

```diff
@@ -105,20 +107,64 @@ final class EphemerisParserSpec extends CatsSuite {
       (e.toMap.from(tail.firstKey) == tail)
     )
   }
-}
 
-object EphemerisParserSpec {
-  private val TimeFormat = DateTimeFormatter.ofPattern("yyyy-MMM-dd HH:mm:ss.SSS")
+  test("Must stream mars") {
+    val head = eph(
+      "2017-Aug-01 00:00:00.000" -> "08 39 26.1113 +19 33 27.498",
+      "2017-Aug-02 20:09:36.000" -> "08 44 16.7664 +19 15 25.619",
+      "2017-Aug-04 16:19:12.000" -> "08 49 06.2770 +18 56 56.576"
+    )
 
-  private def load(n: String): String =
-    Source.fromInputStream(getClass.getResourceAsStream(s"$n.eph")).mkString
+    val s = stream("mars").through(EphemerisParser.elements[IO])
+    val m = TreeMap(s.take(head.size.toLong).runLog.unsafeRunSync: _*)
+
+    assert(m == head)
+  }
+
+  test("Must handle errors") {
+    val z = InstantMicros.ofEpochMilli(0L) -> Coordinates.Zero
+    val s = stream("mars-error")
+             .through(EphemerisParser.elements[IO])
+             .onError(_ => Stream(z))
+    assert(Vector(Some(z)) == s.last.runLog.unsafeRunSync)
+  }
+
+  test("Must stream eitherElements") {
+    val head = Vector[Either[String, Ephemeris.Element]](
+      Right(time("2017-Aug-01 00:00:00.000") -> coords("08 39 26.1113 +19 33 27.498")),
+      Right(time("2017-Aug-02 20:09:36.000") -> coords("08 44 16.7664 +19 15 25.619")),
+      Left("Failure reading:solarPresence")
+    )
+
+    val s = stream("mars-error").through(EphemerisParser.eitherElements[IO])
+    val m = s.take(head.size.toLong).runLog.unsafeRunSync()
+
+    assert(m == head)
+  }
+
+  test("Must stream validElements") {
+    val head = eph(
+      "2017-Aug-01 00:00:00.000" -> "08 39 26.1113 +19 33 27.498", // 0
+      "2017-Aug-02 20:09:36.000" -> "08 44 16.7664 +19 15 25.619", // 1
+      "2017-Aug-06 12:28:48.000" -> "08 53 54.4942 +18 38 00.694"  // 3 (skipping 2)
+    )
 
-  private def time(s: String): InstantMicros =
-    InstantMicros.truncate(LocalDateTime.parse(s, TimeFormat).toInstant(ZoneOffset.UTC))
+    val s = stream("mars-error").through(EphemerisParser.validElements[IO])
+    val m = TreeMap(s.take(head.size.toLong).runLog.unsafeRunSync: _*)
 
-  private def coords(s: String): Coordinates =
-    Coordinates.parse(s).getOrElse(Coordinates.Zero)
+    assert(m == head)
+  }
+
+}
 
-  private def eph(elems: (String, String)*): TreeMap[InstantMicros, Coordinates] =
-    TreeMap(elems.map { case (i, c) => time(i) -> coords(c) }: _*)
+object EphemerisParserSpec {
+  private def inputStream(n: String): InputStream =
+    getClass.getResourceAsStream(s"$n.eph")
+
+  private def stream(n: String): Stream[IO, String] =
+    fs2.io.readInputStream(IO(inputStream(n)), 128)
+          .through(fs2.text.utf8Decode)
```

Comment:
Something curious is going on here.  The tests are sensitive to the value for  the buffer size.  Usually you'd want a value of 8K or something, but I thought it would be good to make sure that the entire file doesn't fit in a single chunk.  If the buffer size is below 110 bytes, everything inexplicably fails.

Reply:
Hmm, weird. Can you minimize this failure?